lib/autoupdate/progress.go

```diff
@@ -0,0 +1,81 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package autoupdate
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strings"
+)
+
+var (
+	// ErrCanceled represent the cancellation error for Ctrl-Break/Ctrl-C, depends on the platform.
+	ErrCanceled = fmt.Errorf("canceled")
+)
+
+// cancelableTeeReader is a copy of TeeReader with ability to react on signal notifier
+// to cancel reading process.
+func cancelableTeeReader(r io.Reader, w io.Writer, signals ...os.Signal) io.Reader {
```

Comment:
It's safe to call Close while concurrently Reading from a response body. Instead of creating this custom Reader, I'd recommend something like this:
```golang
go func() {
   <-ctx.Done()
   resp.Body.Close()
}()
```

Reply:
Actually, NewRequestWithContext should cover reading the response body. Have you tested without using cancelableTeeReader? Seems like it should just work.

Reply:
I've replaced to `NewRequestWithContext` recently, let me check